src/xml_threat_model_reviewer.py

```diff
@@ -0,0 +1,115 @@
+                            self.no_boundary_nodes = threat_model.get_no_threat_boundary_node_names()
+
+class XMLThreatModelImageAddToMessageCapability(AgentCapability, ThreatModelDataExtractor):
+    def __init__(self, context: TurnContext, say_when_evaluating: bool, max_width: int, **kwargs):
+        self.say_when_evaluating = say_when_evaluating
+        self.context = context
+        self.max_width = max_width
+        self.img = None
+        
+        super().__init__()
+        super(AgentCapability, self).__init__(**kwargs)
+    
+    def add_to_agent(self, agent: ConversableAgent):
+        agent.register_hook("process_all_messages_before_reply", self._add_data_to_messages)
+        
+    def _add_data_to_messages(self, messages):
+        if self.img is None:
+            self.extract_image_from_state()
+            self.extract_data_from_state()
+            if self.img:
+                jpeg = self.convert_to_jpeg_if_needed(self.img)
+                if jpeg:
+                    # Unfortunately autogen currently doesn't support async nested chats.
+                    # So we need to do this "fire and forget" hack to send the image.
+                    ensure_future(self._say_when_evaluating(jpeg))
+                self.resize(self.max_width)
+        if self.label_names is not None or self.no_boundary_nodes is not None:
+            messages = messages.copy()
+            content = f"""The file details for the file you need to validate are: 
+            1. The nodes with no boundaries are {self.no_boundary_nodes}.
+            2. The list of label names is {self.label_names}."""
+            messages.append({"content": content, "role": "user"})
+        else:
+            messages = messages.copy()
+            messages.append({"content": "No threat model exists.", "role": "user"})
+        return messages
+    
+    def resize(self, max_width: int):
+        assert self.img is not None, "There is no image to resize!"
+        new_img = Image.new(self.img.mode, self.img.size)
+        wpercent = (max_width / float(self.img.size[0]))
+        hsize = int((float(self.img.size[1]) * float(wpercent)))
+        self.img = new_img.resize((max_width, hsize))
+    
+    async def _say_when_evaluating(self, img: Image.Image):
+        if self.say_when_evaluating:
+            jpeg = self.convert_to_jpeg_if_needed(img)
+            if jpeg:
+                await self.context.send_activity(
+                    Activity(
+                        type=ActivityTypes.message,
+                        text="Here is the threat model we are evaluating",
+                        attachments=[Attachment(content_type="image/jpeg", content_url=pil_to_data_uri(jpeg))]
+                    )
+                )
+
+def setup_xml_threat_model_reviewer(llm_config, context: TurnContext, state: AppTurnState, threat_model_spec: str = """
+    1. All nodes should be inside a boundary. Are there any nodes not in a boundary?
```

Comment:
NIT: the indentations will all go out as whitespace in the context. Is that just an artifact of the code editor or intentional?

Reply:
Not intentional. I'll update.